yearly_lusmatrix.R

```diff
@@ -0,0 +1,151 @@
+#DYNAMIC LUSMATRIX XLS FILE TO YEARLY LUSMATRIX TXT FILES
+
+#Created by Bep Schrammeijer
+#July 2016
+#Creates lusmatrix files for each year
+
+#packages required
+require(openxlsx)
+
+#load excel workbook ls_matrix_country_ls24_origIMAGEd from.../Scenario_SSP1/IME/ME/ME/lus_matrix_ME_ls24.xlsx, call it wb
+ls_matrix <- loadWorkbook("Scenario_SSP1/IME/ME/ME/lus_matrix_ME_ls24.xlsx")
+#year_0
+ls_matrix_0 <- readWorkbook("Scenario_SSP1/IME/ME/ME/lus_matrix_ME_ls24.xlsx", sheet = "dynamic output", 16, colNames = FALSE, rowNames = FALSE, rows = 16:39, cols = 2:5)
+ls_matrix_0[,1:3] <- round(ls_matrix_0[,1:3], 0)
+ls_matrix_0[,4] <- round(ls_matrix_0[,4], 2)
+write.table(ls_matrix_0, "Scenario_SSP1/IME/ME/ME/lusmatrix.0", sep = "\t", row.names = FALSE, col.names = FALSE)
```

Comment:
As far I can understand, there's really only one thing you want to do with the only thing changing being the year. So, the following things are the only ones changing:

1. Region of columns (argument `cols`) read in
2. The name of of the output file (reflecting the year number)

We're in luck! This is a perfect opportunity to [refactor](https://en.wikipedia.org/wiki/Code_refactoring) the code into a [function.](http://adv-r.had.co.nz/Functions.html). Here's an alternative implementation as a function:

```r
process_year <- function(woorkbook, sheet_name, rows, cols,
                         output_dir) {
  # Read in the data. Pass on the function deafault arguments, leave
  # others hard coded (e.g. col and row names)
  ls_matrix <- readWorkbook(woorkbook, sheet = sheet_name,
                            colNames = FALSE, rowNames = FALSE,
                            rows = rows, cols = cols)
  # Round the columns
  ls_matrix[,1:3] <- round(ls_matrix[,1:3], 0)
  ls_matrix[,4] <- round(ls_matrix[,4], 2)
  # Dynamically generate the file name. We could provide the number
  # of the year as an argument, but we can also read it from the
  # Excel file since it's always in the same place. Use the first
  # elements of the rows and cols vectors.
  year_no_row <- rows[1] - 3
  year_no_col <- cols[1]
  year_no <- readWorkbook(woorkbook, sheet = sheet_name,
                          colNames = FALSE, rowNames = FALSE,
                          rows = year_no_row, cols = year_no_col)
  # Object year_no is still a data.frame, let's extract the first
  # (and only) atomic element to get a single number. Note
  # that we are overwriting the previous object.
  year_no <- year_no[1, 1]
  # Let's give some info for the user as well
  message("Processing year ", year_no, "...")
  # Construct the file name. Use file.path to create the path
  # in a platform independent way.
  output_file <- file.path(output_dir, paste0("lusmatrix.", year_no))
  write.table(ls_matrix, output_file, sep = "\t",
              row.names = FALSE, col.names = FALSE)

  # No need to worry about the exact meaning of this, but it's good
  # to return something.
  return(invisible(TRUE))
}
```